lib/Lift/InstructionTranslator.cpp

```diff
@@ -733,47 +626,72 @@ IT::TranslationResult IT::translateCall(PTCInstruction *Instr) {
                                        ArrayRef<Type *>(InArgsType),
                                        false);
 
-  std::string HelperName = "helper_" + TheCall.helperName();
+  LibTcgHelperInfo Info = LibTcg.get_helper_info(Instr);
+  std::string HelperName = "helper_" + std::string(Info.func_name);
   FunctionCallee FDecl = TheModule.getOrInsertFunction(HelperName, CalleeType);
 
   FunctionTags::Helper.addTo(cast<Function>(skipCasts(FDecl.getCallee())));
 
   CallInst *Result = Builder.CreateCall(FDecl, InArgs);
 
-  if (TheCall.OutArguments.size() != 0)
+  if (Instr->nb_oargs != 0) {
     Builder.CreateStore(Result, ResultDestination);
+  }
+
+  if (Info.func_flags & LIBTCG_CALL_NO_RETURN) {
+    handleExitTB();
+  }
 
   return Success;
 }
 
 IT::TranslationResult
-IT::translate(PTCInstruction *Instr, MetaAddress PC, MetaAddress NextPC) {
-  const PTC::Instruction TheInstruction(Instr);
-
+IT::translate(LibTcgInstruction *Instr, MetaAddress PC, MetaAddress NextPC) {
   std::vector<Value *> InArgs;
-  for (uint64_t TemporaryId : TheInstruction.InArguments) {
-    auto *Load = Variables.load(Builder, TemporaryId);
+  for (unsigned I = 0; I < Instr->nb_iargs; ++I) {
+    auto *Load = Variables.load(Builder, &Instr->input_args[I]);
     if (Load == nullptr)
       return Abort;
     InArgs.push_back(Load);
   }
 
-  auto ConstArgs = TheInstruction.ConstArguments;
+  // TODO(anjo): Remove this, constant args to eg add are not passed as const but const temps
```

Comment:
Can you elaborate?

Reply:
History, 

a previous version of libtcg (QEMU 6.0?) stored immediate values to instructions in the block of `LibTcgInstruction::constant_args[]`. What the block of code below this line _did_ was separate out immediate values from other constants such as labels/comparison ops/memory ops. Immediate values would go in `InArgs`, and other constants into `ConstArgs`.

But this changed somewhere between 6.0 and 8.2.1, and now immediate values are stored on the TCG Temp struct instead (LibTcgTemp::val), and so the path pushing to `InArgs` is _rarely_ taken. There are still a few cases, op_extract2 and op_st have some argument there at the very least, but the distinction is not necessary anymore, as it's not as common.

So instead we can simply pass in `Instr->constant_args` directly, maybe wrapped in some fancy C++ std::array, and the _few_ instructions which need constants can access them directly.